src/globus_cli/login_manager/tokenstore.py

```diff
@@ -10,15 +11,218 @@
-class _TokenStoreFuncProto:
-    _instance: SQLiteAdapter
+class CLITokenstorage:
+    """
+    A wrapper over the globus-sdk's v1 tokenstorage which provides simplified
+    capabilities specific to the CLI.
+    """
+
+    def __init__(self) -> None:
+        self._adapter: SQLiteAdapter | None = self._construct_adapter()
+
+    def _construct_adapter(self) -> SQLiteAdapter:
+        from globus_sdk.tokenstorage import SQLiteAdapter  # noqa: F811
+
+        # when initializing the token storage adapter, check if the storage file exists
+        # if it does not, then use this as a flag to clean the old config
+        fname = _get_storage_filename()
+        if not os.path.exists(fname):
+            from ._old_config import invalidate_old_config
+
+            invalidate_old_config(self.internal_native_client)
+
+        return SQLiteAdapter(fname, namespace=_resolve_namespace())
+
+    def close(self) -> None:
+        if self._adapter is not None:
+            self._adapter.close()
+            self._adapter = None
+
+    @functools.cached_property
+    def internal_native_client(self) -> globus_sdk.NativeAppAuthClient:
+        """
+        This is the client that represents the CLI itself (prior to templating).
+        """
+        template_id = _template_client_id()
+        return globus_sdk.NativeAppAuthClient(
+            template_id, app_name="Globus CLI (native client)"
+        )
+
+    def internal_auth_client(self) -> globus_sdk.ConfidentialAppAuthClient:
+        """
+        Pull template client credentials from storage and use them to create a
+        ConfidentialAppAuthClient.
+        In the event that credentials are not found, template a new client via the
+        Auth API, save the credentials for that client, and then build and return the
+        ConfidentialAppAuthClient.
+        """
+        if is_client_login():
+            raise ValueError("client logins shouldn't create internal auth clients")
+
+        client_data = self.read_well_known_config("auth_client_data")
+        if client_data is not None:
+            client_id = client_data["client_id"]
+            client_secret = client_data["client_secret"]
+        else:
+            # register a new instance client with auth
+            nc = self.internal_native_client
+            res = nc.post(
+                "/v2/api/clients",
+                data={"client": {"template_id": nc.client_id, "name": "Globus CLI"}},
+            )
+            # get values and write to config
+            credential_data = res["included"]["client_credential"]
+            client_id = credential_data["client"]
+            client_secret = credential_data["secret"]
+
+            self.store_well_known_config(
+                "auth_client_data",
+                {"client_id": client_id, "client_secret": client_secret},
+            )
+
+        return globus_sdk.ConfidentialAppAuthClient(
+            client_id, client_secret, app_name="Globus CLI"
+        )
```

Comment:
Is it worth thinking about caching this value?

I realize we want to do the `is_client_login` check, but could still throw it in an instance variable and return it if it's already been instantiated.

Reply:
Yeah, simply wrapping this in `cached_property` poses no issues, as far as I can tell (one mock needed a tweak, that was all). I'm including it.

Reply:
Do we need to evict the cache if someone calls `delete_templated_client`?

Reply:
I don't think it's necessary (although it would be more conceptually clean). It can only happen during logout when a user uses the (hidden) `--delete-client` flag, and the client is never used after that in that process.

Would you like me to add that? I believe it's just a matter of doing a `del self.cli_confidential_client` at the end of the delete.